src/pkg/cli/client/byoc/aws/domain.go

```diff
@@ -0,0 +1,115 @@
+package aws
+
+import (
+	"context"
+	"errors"
+	"slices"
+	"strings"
+
+	"github.com/DefangLabs/defang/src/pkg/clouds/aws"
+	"github.com/DefangLabs/defang/src/pkg/dns"
+	"github.com/DefangLabs/defang/src/pkg/term"
+	"github.com/DefangLabs/defang/src/pkg/track"
+	"github.com/aws/aws-sdk-go-v2/service/route53/types"
+)
+
+func prepareDomainDelegation(ctx context.Context, projectDomain string, r53Client aws.Route53API) (nsServers []string, delegationSetId string, err error) {
+	// There's four cases to consider:
+	//  1. The subdomain zone does not exist: we create/get a delegation set and get its NS records and let CD/Pulumi create the hosted zone
+	//  2. The subdomain zone exists:
+	//    a. The zone was created by the older CLI: we need to get the NS records from the existing zone and pass to Fabric; no delegation set
+	//    b. The zone was created by the new CD/Pulumi: we get the NS records from the delegation set and let CD/Pulumi create/update the hosted zone
+	//    c. The zone was created another way: get the NS records from the existing zone and pass to Fabric; no delegation set
+
+	var delegationSet *types.DelegationSet
+	zone, err := aws.GetHostedZoneByName(ctx, projectDomain, r53Client)
+	if err != nil {
+		// The only acceptable error is that the zone was not found
+		if !errors.Is(err, aws.ErrZoneNotFound) {
+			return nil, "", err // TODO: we should not fail deployment if GetHostedZoneByName fails
+		}
+		term.Debugf("Zone %q not found, delegation set will be created", projectDomain)
+
+		// Case 1: The zone doesn't exist: we'll create/get a delegation set and let CD/Pulumi create the hosted zone
+		delegationSet, err = prepareDomainDelegationFromDelegationSet(ctx, r53Client)
+		if err != nil {
+			return nil, "", err
+		}
+	} else {
+		// Case 2: Get the NS records for the existing subdomain zone
+		delegationSet, err = prepareDomainDelegationFromZone(ctx, zone, r53Client)
+		if err != nil {
+			return nil, "", err
+		}
+	}
+
+	if len(delegationSet.NameServers) == 0 {
+		return nil, "", errors.New("no NS records found for the delegation set") // should not happen
+	}
+	if delegationSet.Id != nil {
+		term.Debug("Route53 delegation set ID:", *delegationSet.Id)
+		delegationSetId = strings.TrimPrefix(*delegationSet.Id, "/delegationset/")
+	}
+
+	return delegationSet.NameServers, delegationSetId, nil
+}
+
+func prepareDomainDelegationFromDelegationSet(ctx context.Context, r53Client aws.Route53API) (*types.DelegationSet, error) {
```

Comment:
tempted to call this `findOrCreateDelegationSet`. The current name suggests that there should be a `delegationSet` parameter, and there isn't one.